### ISSUE TITLE
`Types::Constrained` behaves inconsistently

### ISSUE THREAD

**Issue by v-kolesnikov** [dry-rb/dry-types#269]
`Dry::Types::<Primitive>` does not check the actual type of input value, but the same type with any constraint will check it, so it's unexpected.

The some examples:

```ruby
# frozen_string_literal: true

require 'bundler/inline'

gemfile(true) do
  gem 'dry-types', '0.13.2'
  gem 'rspec'
end

module Types
  include Dry::Types.module
end

describe Dry::Types do
  it { expect(Types::Float[1]).to eql 1 }
  it { expect(Types::Float[1.0]).to eql 1.0 }

  it { expect(Types::Float.constrained(gt: 0.0)[1]).to eql 1 }
  it { expect(Types::Float.constrained(gt: 0.0)[1.0]).to eql 1.0 }
end
```

```
Failures:

  1) Dry::Types
     Failure/Error: it { expect(Types::Float.constrained(gt: 0.0)[1]).to eql 1 }

     Dry::Types::ConstraintError:
       1 violates constraints (1 must be an instance of Float failed)
     # /Users/v.kolesnikov/.rvm/gems/ruby-2.5.0/gems/dry-types-0.13.2/lib/dry/types/constrained.rb:28:in `block in call'
     # /Users/v.kolesnikov/.rvm/gems/ruby-2.5.0/gems/dry-types-0.13.2/lib/dry/types/definition.rb:74:in `try'
     # /Users/v.kolesnikov/.rvm/gems/ruby-2.5.0/gems/dry-types-0.13.2/lib/dry/types/constrained.rb:43:in `try'
     # /Users/v.kolesnikov/.rvm/gems/ruby-2.5.0/gems/dry-types-0.13.2/lib/dry/types/constrained.rb:27:in `call'
     # ./issues/dry_types_constrained_spec.rb:18:in `block (2 levels) in <top (required)>'

Finished in 0.00749 seconds (files took 1.46 seconds to load)
4 examples, 1 failure
```

[Discussion](https://gitter.im/dry-rb/chat?at=5b6afb9fcf8ab4758abcc7f3)

**Comment by solnic** [comment]
For the record, this happens because in this case `Constrained` uses `Definition#try` and it [performs `valid?` check](https://github.com/dry-rb/dry-types/blob/master/lib/dry/types/definition.rb#L71), which results in a failure.

One option to fix this is to make `Constrained#try` check if its underlying type is not a plain definition [here](https://github.com/dry-rb/dry-types/blob/master/lib/dry/types/constrained.rb#L44) and only then use try, otherwise just return the result.

**Comment by flash-gordon** [comment]
This is addressed per #309
